File: ArtSourceWrapper/Flickr.cs
using FlickrNet;
using System;
using System.Collections.Generic;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace ArtSourceWrapper {
    public class FlickrWrapper : SiteWrapper<FlickrSubmissionWrapper, int> {
        private Flickr _flickr;

        public FlickrWrapper(string apiKey, string sharedSecret, string oAuthAccessToken, string oAuthAccessTokenSecret) {
            _flickr = new Flickr(apiKey, sharedSecret) {
                OAuthAccessToken = oAuthAccessToken,
                OAuthAccessTokenSecret = oAuthAccessTokenSecret
            };
        }

        public FlickrWrapper(Flickr client) {
            _flickr = client;
        }

        public override string SiteName => "Flickr";
        public override string WrapperName => "Flickr";

        public override int BatchSize { get; set; } = 100;
        public override int MinBatchSize => 1;
        public override int MaxBatchSize => 500;

        private FoundUser _cachedUser;

        public async Task<FoundUser> GetUserAsync() {
            if (_cachedUser == null) {
                var t = new TaskCompletionSource<Auth>();
                _flickr.AuthOAuthCheckTokenAsync(result => {
                    if (result.HasError) {
                        t.SetException(result.Error);
                    } else {
                        t.SetResult(result.Result);
                    }
                });
                var oauth = await t.Task;
                _cachedUser = oauth.User;
            }
            return _cachedUser;
        }
        
        public async override Task<string> WhoamiAsync() {
            return (await GetUserAsync()).UserName;
        }

        public async override Task<string> GetUserIconAsync(int size) {
            string id = (await GetUserAsync()).UserId;

            var t = new TaskCompletionSource<Person>();
            _flickr.PeopleGetInfoAsync(id, result => {
                if (result.HasError) {
                    t.SetException(result.Error);
                } else {
                    t.SetResult(result.Result);
                }
            });
            var person = await t.Task;

            return person.IconServer != "0"
                ? $"http://farm{person.IconFarm}.staticflickr.com/{person.IconServer}/buddyicons/{person.UserId}.jpg"
                : $"https://www.flickr.com/images/buddyicon.gif";
        }

        protected async override Task<InternalFetchResult> InternalFetchAsync(int? startPosition, int count) {
            var t = new TaskCompletionSource<PhotoCollection>();
            _flickr.PeopleGetPhotosAsync("me",
                PhotoSearchExtras.Description | PhotoSearchExtras.Tags | PhotoSearchExtras.DateUploaded | PhotoSearchExtras.OriginalFormat,
                startPosition ?? 1,
                count,
                result => {
                    if (result.HasError) {
                        t.SetException(result.Error);
                    } else {
                        t.SetResult(result.Result);
                    }
                });
            var r = await t.Task;

            return new InternalFetchResult(r.Select(p => new FlickrSubmissionWrapper(p)), r.Page + 1, r.Page == r.Pages);
        }
    }

    public class FlickrSubmissionWrapper : ISubmissionWrapper {
        private Photo _photo;
        public FlickrSubmissionWrapper(Photo photo) {
            _photo = photo;
        }

        public string Title => _photo.Title;
        public string HTMLDescription => _photo.Description;
        public bool Mature => false;
        public bool Adult => false;
		public IEnumerable<string> Tags => _photo.Tags;
        public DateTime Timestamp => _photo.DateUploaded;
        public string ViewURL => $"https://www.flickr.com/photos/{_photo.UserId}/{_photo.PhotoId}";
        public string ImageURL => $"https://farm{_photo.Farm}.staticflickr.com/{_photo.Server}/{_photo.PhotoId}_{_photo.OriginalSecret}_o.{_photo.OriginalFormat}";
        public string ThumbnailURL => $"https://farm{_photo.Farm}.staticflickr.com/{_photo.Server}/{_photo.PhotoId}_{_photo.Secret}_q.jpg";
        public Color? BorderColor => null;
    }
}

File: SourceWrappers/Flickr.fs
namespace SourceWrappers

open FlickrNet
open System.Threading.Tasks
open FSharp.Control

type FlickrPostWrapper(photo: Photo) =
    interface IRemotePhotoPost with
        member this.Title = photo.Title
        member this.HTMLDescription = photo.Description
        member this.Mature = false
        member this.Adult = false
        member this.Tags = photo.Tags :> seq<string>
        member this.Timestamp = photo.DateUploaded
        member this.ViewURL = sprintf "https://www.flickr.com/photos/%s/%s" photo.UserId photo.PhotoId
        member this.ImageURL = sprintf "https://farm%s.staticflickr.com/%s/%s_%s_o.%s" photo.Farm photo.Server photo.PhotoId photo.OriginalSecret photo.OriginalFormat
        member this.ThumbnailURL = sprintf "https://farm%s.staticflickr.com/%s/%s_%s_q.jpg" photo.Farm photo.Server photo.PhotoId photo.Secret

type FlickrSourceWrapper(flickr: Flickr) =
    inherit AsyncSeqWrapper()

    let flickrCall f1 = async {
        let tcs = new TaskCompletionSource<'a>()
        f1(new System.Action<FlickrResult<'a>>(fun r -> if r.HasError then tcs.SetException r.Error else tcs.SetResult r.Result))
        return! tcs.Task |> Async.AwaitTask
    }

    let mutable user: FoundUser = null

    let getUser = async {
        if isNull user then
            let! oauth = flickrCall (fun callback -> flickr.AuthOAuthCheckTokenAsync(callback))
            user <- oauth.User
        return user
    }

    new(apiKey: string, sharedSecret: string, oAuthAccessToken: string, oAuthAccessTokenSecret: string) =
        FlickrSourceWrapper(new Flickr(apiKey, sharedSecret, OAuthAccessToken = oAuthAccessToken, OAuthAccessTokenSecret = oAuthAccessTokenSecret))

    override this.Name = "Flickr"

    override this.StartNew() = asyncSeq {
        let mutable cursor = 1
        let mutable more = true
        let extras = PhotoSearchExtras.Description ||| PhotoSearchExtras.Tags ||| PhotoSearchExtras.DateUploaded ||| PhotoSearchExtras.OriginalFormat
        
        while more do
            let! posts = flickrCall (fun callback -> flickr.PeopleGetPhotosAsync("me", extras, cursor, 100 |> min 500, callback))
            for p in posts do
                yield new FlickrPostWrapper(p) :> IPostBase
            
            cursor <- posts.Page + 1
            more <- posts.Page < posts.Pages
    }

    override this.Whoami = async {
        let! user = getUser
        return user.UserName
    }

    override this.GetUserIcon size = async {
        let! user = getUser
        
        let! person = flickrCall (fun callback -> flickr.PeopleGetInfoAsync(user.UserId, callback))
        return if person.IconServer = "0"
            then "https://www.flickr.com/images/buddyicon.gif"
            else sprintf "http://farm%s.staticflickr.com/%s/buddyicons/%s.jpg" person.IconFarm person.IconServer person.UserId
    }